feat: Implement LayeredTracing

mkmik opened the pull request:
__Rationale__

We currently use the `tracing` framework to output to both log outputs (e.g. stdout for k8s) and distributed tracing collectors (e.g. opentelemetry jaeger).

However, due to a limitation in the `tracing` SDK, we can only have one "filter" level that applies
to both logs and tracing outputs. Because of this limitation, currently, distributed trace export and good old logging are mutually exclusive (i.e. as soon as tracing exporter is enabled (e.g. `--traces-exporter=jaeger`) iox disables emitting logs to stdout).

Distributed tracing and local logging have both their strengths, and we should be able to use both at the same time.

This PR adds a `FilteredLayer` tracing subscriber layer, that wraps a subscriber layer with a independent
filter. We can thus have multiple tracing subscribers, each with a fully independent filter config (we can already control verbosity level for each module).

Distributed tracing subscribers also apply a (configurable) sampling policy, which is not suitable for logging. This allows for enabling a very high verbosity for a very small sample of spans (or for an explicitly marked conversation).

__Note__

This PR just implements the `FilteredLayer` and a test. Another PR will integrate this with
our log/tracing setup code.

tustvold reviewed (commented):
Why do we want trace verbosity level for distributed tracing? I fear this will enable a large number of callsites with significant performance implications?

I'd imagine we'd use `info` spans for the stuff we'd want sent to jaeger? And possibly strip out everything below info in release builds (I believe this is what linkerd-proxy does). See [here](https://docs.rs/tracing/0.1.26/tracing/level_filters/index.html#compile-time-filters)

alamb commented on Cargo.toml:
```diff
@@ -70,6 +70,7 @@ bytes = "1.0"
 chrono = "0.4"
 clap = "2.33.1"
 csv = "1.1"
+delegate = "0.5.1"
```
oh no! More packages! (this brings in 2 new dependencies it seems)

alamb commented on src/commands/layered_tracing.rs:
```diff
@@ -0,0 +1,138 @@
+            self.inner.on_event(event, ctx)
+        }
+    }
+    delegate! {
```
I suggest we do some small copy/paste transgression here to avoid the dependency (and even longer compile times)

mkmik replied:
ok, I actually started with manually delegating code then "cleaned up" with the macro :-)

mkmik replied:
done

alamb replied:
thank you

alamb commented on Cargo.toml:
```diff
@@ -112,6 +113,7 @@ entry = { path = "entry" }
 influxdb2_client = { path = "influxdb2_client" }
 influxdb_iox_client = { path = "influxdb_iox_client", features = ["flight"] }
 test_helpers = { path = "test_helpers" }
+tracing-appender = "0.1.2"
```
Is this one used at all? I didn't see it referenced anywhere in the code

mkmik replied:
yes, it's used in the tests

```
 let (log_file_writer, _log_guard) = tracing_appender::non_blocking(log_file);
```

opentelemetry is quite complicated. You don't pass a "writer" to it, no no; you pass a writer factory (MakeWriter).
So you cannot just pass a `&Vec<[u8]>` and call it a day; you have to plumb a shared Writer into something that can be returned multiple times by a `Fn` ... that's what `tracing_appender` does.

alamb replied:
![image](https://user-images.githubusercontent.com/490673/121603081-3c053e80-ca16-11eb-83dd-47c5ee58dcb6.png)

mkmik replied:
ok, I traded the `tracing_appender` for `synchronized-writer` to make the test actually deterministic.

alamb commented on src/commands/layered_tracing.rs:
```diff
@@ -0,0 +1,138 @@
+
+    #[test]
+    fn test_filtered_layer() {
+        let (log_file, log_file_path) = NamedTempFile::new().expect("temp file").into_parts();
```
This is a very cool test --send info/error to one stream and trace/info/error to another

alamb reviewed (approved):
I think this looks good -- I think the fact there is a difference between `tracing`  and the level `trace!` makes this conversation quite confusing.

Having the ability for one stream to emit at one "level" and then tap a second stream with a more restrictive filter makes all sorts of sense to me 👍

jacobmarble commented on src/commands/layered_tracing.rs:
```diff
@@ -0,0 +1,145 @@
+
+        let subscriber = tracing_subscriber::fmt()
+            .json()
+            .with_env_filter(EnvFilter::new("debug"))
+            .with_writer(trace_file_writer)
+            .without_time()
+            .with_target(false)
+            .finish()
+            .with(FilteredLayer::new(
+                EnvFilter::new("info"),
```
When I swap the values "info" and "debug", the test fails only at line 136, but I would expect line 128 to also fail. This suggests that filters are not both receiving the same stream of trace events, but that one is filtering the other.

mkmik replied:
yes, that's how this particular test has been set up; if you want two completely independent subscribers you need to use two `FilteredLayer` (which is likely how this will be used in `tracing.rs`)

Do you think I should adapt the test to showcase that?

jacobmarble replied:
Yes, that would illustrate the goal that this is trying to reach.

mkmik replied:
done PTAL

mkmik commented on src/commands/layered_tracing.rs:
```diff
@@ -0,0 +1,145 @@
+        }
+    }
+
+    fn new_span(&self, attrs: &Attributes<'_>, id: &Id, ctx: Context<'_, S>) {
+        self.inner.new_span(attrs, id, ctx)
+    }
+    fn on_record(&self, span: &Id, values: &Record<'_>, ctx: Context<'_, S>) {
+        self.inner.on_record(span, values, ctx)
+    }
+    fn on_follows_from(&self, span: &Id, follows: &Id, ctx: Context<'_, S>) {
+        self.inner.on_follows_from(span, follows, ctx)
+    }
+    fn on_enter(&self, id: &Id, ctx: Context<'_, S>) {
+        self.inner.on_enter(id, ctx)
+    }
+    fn on_exit(&self, id: &Id, ctx: Context<'_, S>) {
+        self.inner.on_exit(id, ctx)
+    }
+    fn on_close(&self, id: Id, ctx: Context<'_, S>) {
+        self.inner.on_close(id, ctx)
+    }
+    fn on_id_change(&self, old: &Id, new: &Id, ctx: Context<'_, S>) {
+        self.inner.on_id_change(old, new, ctx)
+    }
```
@jacobmarble btw, I noticed that these functions are not even called in my tests.

Not sure if we really need to delegate enything other than `on_event`. I'm too lazy right now to actually read all the docs and understand what those callbacks are for. Do you know?

jacobmarble replied:
For delegating, the `inner` implementation should filter its own `on_event`, so it could be simply delegated.

mkmik replied:
can you please rephrase, I don't understand

jacobmarble replied:
Yes. Any implementation of `Layer.on_event` should filter its input. This covers situations where the callback is registered as "maybe" so that the Layer can decide whether to filter when the event is emitted.

So, just delegate it like the other methods.

mkmik replied:
I'm still confused; are you suggesting that we should chain the filter onto `inner` and then just delegate everything except `register_callsite` and `enabled` ?

mkmik replied:
An `on_event` implementation doesn't have to do it's own filtering. `tracing` won't call `on_event` if the subscriber declares its not interested in the event either by returning `Interest::never()` from `register_callsite` or by returning `false` to the call to `enabled`.

The wrapper has to do its own filtering.

jacobmarble commented:
Related https://github.com/tokio-rs/tracing/issues/597

mkmik commented:
> Related [tokio-rs/tracing#597](https://github.com/tokio-rs/tracing/issues/597)

Thanks for the link.

One of the solutions there (https://github.com/tokio-rs/tracing/issues/597#issuecomment-814507031) shows how to do the filtering explicitly by metadata level.

This PR instead delegates the checking to an EnvFilter which can apply more complex filters, like applying different levels for different rust modules.

mkmik commented on src/commands/layered_tracing.rs:
```diff
@@ -0,0 +1,178 @@
+    #[allow(dead_code)]
+    pub fn new(filter: EnvFilter, delegate: L) -> Self {
+        Self {
+            inner: delegate.and_then(filter),
```
@jacobmarble I found  this other trick: by chaining the filter to the delegate, I can get rid of the phantomdata and I simplify the call to `enabled`, since now the delegate itself knows when it's enabled.

mkmik commented:
hmm it seems that the test is still flaky despite the calls to `.flush()`

```
---- commands::layered_tracing::tests::test_independent stdout ----
thread 'commands::layered_tracing::tests::test_independent' panicked at 'assertion failed: `(left == right)`
  left: `[]`,
 right: `["{\"timestamp\":\"\",\"level\":\"ERROR\",\"fields\":{\"message\":\"foo\"}}", "{\"timestamp\":\"\",\"level\":\"INFO\",\"fields\":{\"message\":\"bar\"}}"]`', src/commands/layered_tracing.rs:170:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

mkmik commented:
turns out we call `tracing::subscriber::set_global_default(subscriber)?;` during e2e tests, which conflicts with testing the subcribers here

mkmik commented:
ok, the test was failing because the writer I was using used a background thread. Switched to a synchronized writer into a vec<u8> and simplified the test a bit.

e-dard commented:
This is very cool. Not going to review as you have had so much good feedback already 👍
